plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +302,243 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+			gs_plugin_event_set_error (event, local_error);
+			gs_plugin_event_add_flag (event, GS_PLUGIN_EVENT_FLAG_WARNING);
+			gs_plugin_report_event (plugin, event);
+		}
+
+		break;
+	}
+	case EOS_UPDATER_STATE_FETCHING: {
+		guint64 total_size = 0;
+		guint64 downloaded = 0;
+		gfloat progress = 0;
+
+		/* ensure the state transition to 'installing' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+		downloaded = eos_updater_get_downloaded_bytes (priv->updater_proxy);
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
+
+		if (total_size == 0) {
+			g_debug ("OS upgrade %s total size is 0!",
+				 gs_app_get_unique_id (app));
+		} else {
+			/* set progress only up to 90%, leaving the remaining
+			 * for applying the update */
+			progress = (gfloat) downloaded / (gfloat) total_size * 90.0;
+		}
+		gs_app_set_progress (app, (guint) progress);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_READY: {
+		/* if there's an update ready to be deployed and we are not yet
+		 * showing the OS upgrade as downloading (state 'installing'), then
+		 * just show it as available so the user has a change to click 'download'
+		 * which will end up deploying the update */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING) {
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+			break;
+		}
+
+		gs_app_set_progress (app, 95);
+
+		/* if the OS upgrade is being installed, but the updater state is
+		 * 'update-ready', it means we should proceed to applying the upgrade */
+		eos_updater_call_apply (priv->updater_proxy, NULL, NULL, NULL);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_APPLYING_UPDATE: {
+		/* ensure the state transition to 'installing' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+
+		/* set as 'installing' because if it is applying the update, we
+		 * want to show the progress bar */
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+		gs_app_set_progress (app, 95);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_APPLIED: {
+		/* ensure the state transition to 'updatable' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+		gs_app_set_progress (app, 100);
+		gs_app_set_state (app, AS_APP_STATE_UPDATABLE);
+
+		break;
+	}
+	default:
+		break;
+	}
+
+	/* ensure that a version (and URL) are set */
+	version = gs_app_get_version (app);
+	if (version == NULL || *version == '\0') {
+		version = eos_updater_get_version (priv->updater_proxy);
+		gs_app_set_version (app, version);
+	}
+
+	/* reset the 'download-by-user' state if the the app is no longer
+	 * shown as downloading */
+	if (previous_app_state != AS_APP_STATE_INSTALLING)
+		os_upgrade_set_download_by_user (app, FALSE);
```

Comment:
I am not sure the comment and the code agree here; `previous_app_state != AS_APP_STATE_INSTALLING` means that prior to this state change we were not polling, fetching, or installing the update (i.e. we were not showing the app as downloading). Now we may be in one of those states, and we may be showing the app as downloading.

The comment instead suggests that the condition is there to handle the case where we used to show the app as downloading and now we're not anymore, which is the opposite.

(I think that the code is correct and we only want to clear the flag after we stopped downloading/installing.)

Reply:
The `previous_app_state` is wrong here. The `os_upgrade_set_download_by_user` sets a metadata key that tells us that the download has been started by the user. When the app changes to not-downloading, then we need to reset that. Since this is used to decide whether to show warnings on errors, it was still working for that because we do set the installing state before polling/fetching, so the previous state will always be `installing` and the `download-by-user` will keep being `TRUE`.
So the correct state to check is the current (I don't think this is what you said) and I will change it.